EIPS/eip-7923.md

```diff
@@ -0,0 +1,283 @@
+-        if new_size <= len(self):
++        if start_position + size > self.msize:
++            self.msize = ceil32(start_position + size)
++
++    def write(self, start_position: int, size: int, value: bytes) -> None:
++        if size == 0:
+             return
+
+-        size_to_extend = new_size - len(self)
+-        try:
+-            self._bytes.extend(bytearray(size_to_extend))
+-        except BufferError:
+-            # we can't extend the buffer (which might involve relocating it) if a
+-            # memoryview (which stores a pointer into the buffer) has been created by
+-            # read() and not released. Callers of read() will never try to write to the
+-            # buffer so we're not missing anything by making a new buffer and forgetting
+-            # about the old one. We're keeping too much memory around but this is still
+-            # a net savings over having read() return a new bytes() object every time.
+-            self._bytes = self._bytes + bytearray(size_to_extend)
+-
+-    def __len__(self) -> int:
+-        return len(self._bytes)
++        if start_position + size >= 2**32:
++            raise PyEVMError("Non 32-bit address")
+
+-    def write(self, start_position: int, size: int, value: bytes) -> None:
+-        if size:
+-            validate_uint256(start_position)
+-            validate_uint256(size)
+-            validate_is_bytes(value)
+-            validate_length(value, length=size)
+-            validate_lte(start_position + size, maximum=len(self))
++        validate_uint256(start_position)
++        validate_uint256(size)
++        validate_is_bytes(value)
++        validate_length(value, length=size)
++
++        end_position = start_position + size
+
+-            self._bytes[start_position : start_position + len(value)] = value
++        self.memview[start_position : end_position] = value
+
++    # unused
+     def read(self, start_position: int, size: int) -> memoryview:
+-        return memoryview(self._bytes)[start_position : start_position + size]
++        return memoryview(self.memview)[start_position : start_position + size]
+
+     def read_bytes(self, start_position: int, size: int) -> bytes:
+-        return bytes(self._bytes[start_position : start_position + size])
++        return bytes(self.memview[start_position : start_position + size])
+
+     def copy(self, destination: int, source: int, length: int) -> None:
+         if length == 0:
+@@ -69,5 +68,5 @@ class Memory(MemoryAPI):
+         validate_uint256(length)
+         validate_lte(max(destination, source) + length, maximum=len(self))
+
+-        buf = memoryview(self._bytes)
++        buf = memoryview(self.memview)
+         buf[destination : destination + length] = buf[source : source + length]
+```
+
+## Security Considerations
+
+There are two primary security considerations regarding this EIP.
+
+One, does it break existing contracts? That is, do existing contracts depend on memory being restricted?
```

Comment:
This is actually a very interesting one, because I would not find it too exotic that some contracts would try to allocate memory at `2^256-1` and expect this to OOG.

Reply:
in the EIP, 2**32-1 is the max address. i believe in geth, it requires memory pointers to fit in `uint64_t`.

Reply:
Ah sorry, I meant in the practical phase of contracts deployed on mainnet. For instance trying to MSTORE at 0xff..ff to trigger an OOG. This is not unreasonable at all, this would actually be one of my first candidates to write a method which goes OOG without triggering the designated INVALID (0xfe) opcode. 

So with this EIP, that OOG method breaks for values intended to OOG but are lower than this max `2^32-1` :thinking: Since it will not OOG. But it would also be weird to write a method which goes OOG and has this specific `2^32-1` storage location (or lower) and not `2^256-1`.

I don't think there are many cases of this, but we should definitely investigate the impact here.

Reply:
i'm not sure if it helps, but geth has a hard limit for memory addresses of `2**64 - 1` --
https://github.com/ethereum/go-ethereum/blob/9b4eab6a29704f55fa7b4b92e296094f0dbcee22/core/vm/common.go#L26-L33. i assume this is somehow well-known as an implementation-defined limit.

Reply:
i was thinking about this, and i think we should generally have a policy of allowing limits to expand. like another way to generate an exceptional halt is to try to issue a `CREATE` with a buffer of size 24kb+1 ... i don't think that should we never increase the initcode limit, though.